DAY1-5

kyeongsuk-jin opened the pull request.

kyeongsuk-jin committed:
DAY1-5 (#2)

kimchhickey commented on clojure/src/aoc2018_1.clj:
```diff
@@ -1,12 +1,30 @@
+;; 처음으로 두번 나오는 숫자를 리턴하시오.
 ;; 예) +3, +3, +4, -2, -4 는 10이 처음으로 두번 나오는 숫자임.
 ;; 0 -> 3 (+3) -> 6 (+3) -> 10(+4) -> 8(-2) -> 4(-4) -> 7(+3) -> 10(+3) -> ...
+(defn find-second-same-sum [num-arr]
+  (reduce (fn [r x]
+            (let [{:keys [sum-set last-sum]} r
+                          sum (+ x last-sum)]
+                      (if (sum-set sum)
+                        (reduced {:result sum})
+                        {:last-sum sum :sum-set (conj sum-set sum)}))
+            )
+          {:last-sum 0 :sum-set #{}} num-arr))
```
이 부분 `reductions` 코어함수로  리팩토링 해보시겠어여?
이 함수에서 `(+ x last-sum)` 이런 비즈니스 로직을 바깥으로 추출해낼 수 있어서 조금 더 범용적인 함수가 될거에요.

참고 자료로 여기 기술 블로그가 있습니다.
https://green-labs.github.io/clojure-bootcamp

kyeongsuk-jin replied:
수정해서 머지 했습니다.

https://github.com/kyeongsuk-jin/bootcamp_clojure/commit/dd9627bc3bafa5cce3b9f683c4b4a476fdc58fdb

kimchhickey commented on clojure/src/aoc2018_2.clj:
```diff
@@ -13,6 +14,18 @@
+     (map vals)
+     (mapcat set)
+     (frequencies)
+     ((juxt get-val-2 get-val-3))
+     (apply *))
```
좋습니다 👍

kimchhickey commented on clojure/src/aoc2018_2.clj:
```diff
@@ -26,7 +39,43 @@
+(defn create-map [arr]
+  (for [x (range (dec (count arr)))
+        str1 (subvec arr (inc x))]
+    (let [str2 (part2-arr x)]
```
`part2-arr` 이게 외부 값을 참조하고 있는데, 아래쪽에서 같은 값을 넣어주는 것 같아요. 
arr로 바꿔도 되는거죠?

kyeongsuk-jin replied:
ㅎㅎ 이런 실수를... 수정해서 머지했습니다.

kimchhickey commented on clojure/src/aoc2018_2.clj:
```diff
@@ -26,7 +39,43 @@
+
+(->> (create-map part2-arr)
+     (filter check-one-char-diff?)
+     (map :same-str))
```
원하는 값을 다 구해도 좋지만 이곳에 `first`를 쓰면 지연 평가의 이점을 누릴 수 있을 것 같습니다.

kyeongsuk-jin replied:
넵 first 추가했습니다!

https://github.com/kyeongsuk-jin/bootcamp_clojure/commit/9af9405cd589e93c8980ff6aea97edd10a4ed6df

kimchhickey commented on clojure/src/aoc2018_3.clj:
```diff
@@ -23,10 +26,68 @@
 ;; 겹치는 지역의 갯수를 출력하시오. (위의 예시에서는 4)
 
+; rematched 사용
+(defn parse-input [str]
+  "입력 받은 문자열을 Map 으로 변경한다.
 
+  ex) (parse-grid-input \"#1 @ 1,3: 4x4\")
+  ;=> {:id 1, :start-r 1, :start-c 3, :inc-r 4, :inc-c 4}
+  "
+  (let [match-format #"^#(\d+)\s@\s(\d+),(\d+):\s(\d+)x(\d+)"
+        matched (rest (re-matches match-format str))]
+    (when (some? matched)
+      (zipmap [:id :start-r :start-c :inc-r :inc-c]
+              (->> matched (mapv parse-long)))
+      )))
```
이 함수의 타입 시그니쳐를 생각해보면, map 또는 nil인데, AoC 인풋이 안전하니까 별 문제없이 동작하는 것 같습니다.
하지만 실제로 parse 함수를 작성할 때, 실패할 수 있으니까 호출하는 쪽에서 nil 처리가 필요할거에여.

(잘 모르지만) 자바 타입 시그니처로 써보면 이건 `Optional<Map>`을 반환하는 함수랄까요!

kyeongsuk-jin replied:
넵! parse-input 에서 매칭이 안됐을 경우, 기본값을 셋팅하게 수정했습니다.

https://github.com/kyeongsuk-jin/bootcamp_clojure/commit/77bd5b752dc62b4a8ad94c012d9fadb2ce2ebb2a

kimchhickey commented on clojure/src/aoc2018_3.clj:
```diff
@@ -23,10 +26,68 @@
 ;; 위의 예시에서는 ID 3 이 ID 1, 2와 겹치지 않음. 3을 출력.
 ;; 겹치지 않는 영역을 가진 ID를 출력하시오. (문제에서 답이 하나만 나옴을 보장함)
+
+(set/difference (->> merged-grid
+                     (vals)
+                     (apply set/union))
+                (->> merged-grid
+                     (filter filter-conflicted?)
+                     (vals)
+                     (apply set/union)))
```
집합 연산으로 잘 하셨어요 👍

kimchhickey commented on clojure/src/aoc2018_4.clj:
```diff
@@ -26,8 +29,146 @@
+(defn to-date [str]
+  "시간 문자열을 date 형태로 변경한다.
+  "
+  (let [time-pattern "yyyy-MM-dd hh:mm"
+        date-format (SimpleDateFormat. time-pattern)]
+    (.setTimeZone date-format (SimpleTimeZone. 0 "UTC"))
+    (.parse date-format str)))
```
쓰레딩 매크로처럼 java interop에서 `doto` 혹은 `..` 매크로를 한번 이용해보시는 걸 추천합니다!

kyeongsuk-jin replied:
data-format 부분 수정했습니다

https://github.com/kyeongsuk-jin/bootcamp_clojure/commit/e759fa5ac06f9cf8a5c83846a57438518a40927b

kimchhickey commented on clojure/src/aoc2018_4.clj:
```diff
@@ -26,8 +29,146 @@
+        minute (.getMinutes time)
+        guard-id (get-guard-id action)]
+    (if (some? guard-id)
+      (merge m {:shift-guard guard-id})
```
이미 만들어진 map을 명시적으로 merge할 때가 아니면, assoc이나 update를 더 선호해서 쓰는 편이에요.
한번 바꿔보십쇼!

kyeongsuk-jin replied:
assoc 로 수정했습니다.

https://github.com/kyeongsuk-jin/bootcamp_clojure/commit/e759fa5ac06f9cf8a5c83846a57438518a40927b

kimchhickey commented on clojure/src/aoc2018_4.clj:
```diff
@@ -26,8 +29,146 @@
+        guard-id (get-guard-id action)]
+    (if (some? guard-id)
+      (merge m {:shift-guard guard-id})
+        (case [action]
+          ["falls asleep"] (merge m {:sleep-minute minute})
+          ["wakes up"]
```
`case` 👍

kimchhickey commented on clojure/src/aoc2018_4.clj:
```diff
@@ -26,8 +29,146 @@
+  "parse 후에 바로 reduce 로 {id [total-minute range-frequencies]}"
+  [guard-histories]
+  (-> (reduce
+        (fn [r str]
```
`str` 이 코어 함수 `str`를 덮어쓸 수 있으니까, 지금은 어색하시겠지만 아래 컨벤션에 따라서 s도 괜찮습니다.
https://guide.clojure.style/#idiomatic-names

kyeongsuk-jin replied:
s 로 수정해서 머지했습니다.

kimchhickey commented on clojure/src/aoc2018_4.clj:
```diff
@@ -26,8 +29,146 @@
+  [slept-range]
+  (->> slept-range
+       (reduce
+         (fn [r x]
+           (let [[start end] x]
+             (+ r (- end start))))
+         0)))
```
```suggestion
         (fn [r [start end]]
           (+ r (- end start)))
         0)))
```
이렇게도 되려나요

kimchhickey replied:
그리고 이 함수는 단일 reduce의 wrapper기 때문에, 쓰레딩 매크로를 안써도 될 것 같습니다.

kyeongsuk-jin replied:
넵 let 부분 fn args 로 옮기고 threading macro 제거했습니다.

https://github.com/kyeongsuk-jin/bootcamp_clojure/commit/cb5b30e9d2d8a758bd4851866674c0e4f8159637

kimchhickey commented on clojure/src/aoc2018_4.clj:
```diff
@@ -26,8 +29,146 @@
+    slept-ranges
+    (mapcat (fn [x] (let [[start end] x] (range start end))))
+    (frequencies)
+    )
+  )
```
포메터(cljfmt), 스트럭쳐럴 에디팅을 슬슬 써보시면 괄호를 다루실 때 더 편하실거에요!
Intellij 셋팅이 어려우시면 언제든 말씀 주세요

kyeongsuk-jin replied:
formatting 전체 코드에 적용했습니다.

kimchhickey commented on clojure/src/aoc2018_5.clj:
```diff
@@ -10,7 +11,73 @@
+  (let [chars (->> input string/upper-case distinct vec)]
+    (for [char chars] (merge-removed-alphabet-upper-lower str char))))
+
+(->> (reduced-chars input)
+     (map count)
+     (apply min)
+     )
```
리듀스로 잘 푸셨습니다!
이게 아마 시간이 오래 걸릴텐데, `(time f)`로 감싸서 한번 퍼포먼스를 측정해보시고,
stack 자료구조를 이용해서 리팩토링해보세요!

kyeongsuk-jin replied:
stack 으로 로직 변경했습니다.

https://github.com/kyeongsuk-jin/bootcamp_clojure/commit/2f0e40c33435d7f95d52219891725b84086f577f

kimchhickey reviewed (commented):
2주간의 노력하신 것이 잘 드러나는 것 같습니다.
너무 잘 풀어주셨습니다! 👍

남은 문제도 화이팅입니다!!

kyeongsuk-jin committed:
[REVIEW] use reductions

https://github.com/kyeongsuk-jin/bootcamp_clojure/pull/3

참고 자료 - 기술 블로그
https://green-labs.github.io/clojure-bootcamp

kyeongsuk-jin committed:
[REVIEW] 외부값 참조 수정 및 first 추가

https://github.com/kyeongsuk-jin/bootcamp_clojure/pull/3

kyeongsuk-jin committed:
[REVIEW] parse-input default 값 추가

when 사용시 false 인 경우 nil 이 리턴, parse-input 호출하는 부분에서 filter 로직이 필요함.
-> 필터로직 대신 map 에 default 값을 설정함

https://github.com/kyeongsuk-jin/bootcamp_clojure/pull/3

kyeongsuk-jin committed:
[REVIEW] str 대신 s, 단일 reduce threading macro 제거

str 이 코어 함수 str를 덮어쓸 수 있음 https://guide.clojure.style/#idiomatic-names

https://github.com/kyeongsuk-jin/bootcamp_clojure/pull/3

kyeongsuk-jin committed:
[REVIEW] stack 활용 - peek, pop

https://github.com/kyeongsuk-jin/bootcamp_clojure/pull/3

kyeongsuk-jin committed:
[REVIEW] map merge -> assoc, data-format 수정

https://github.com/kyeongsuk-jin/bootcamp_clojure/pull/3

kyeongsuk-jin committed:
[DAY1-5] day 1-5 수정
